desc/io/hdf5_io.py

```diff
@@ -129,6 +129,11 @@ def read_obj(self, obj, where=None):
                 continue
             if isinstance(loc[attr], h5py.Dataset):
                 s = self._decode_attr(loc, attr)
+                # cast NumPy data types to native Python types
```

Comment:
The main issue is that h5py converts everything to numpy data types when saving, so then when we load the data we need to cast it back to the native python types.

Reply:
I think the best solution to this is declaring the attributes as static like so: https://github.com/PlasmaControl/DESC/blob/5b6bfdce331380867cc912651fc542bf600f0433/desc/magnetic_fields/_core.py#L1086-L1088

That way the actual data type shouldn't matter.

Reply:
@f0uriest Can you explain why this is better? Are you saying we should define all bool and int attributes for every class as static? 

I want to make sure our solution is backwards compatible with old files that were saved with numpy data types, and my hack covers that. 

Also, is the `"_period"` static attribute of the `SplineMagneticField` actually used? I don't see it referenced in the code.

Reply:
so the real issue is actually coming from how the objects are flattened here: https://github.com/PlasmaControl/DESC/blob/5b6bfdce331380867cc912651fc542bf600f0433/desc/io/equilibrium_io.py#L109-L139

along with the default types that we assume are static here: https://github.com/PlasmaControl/DESC/blob/5b6bfdce331380867cc912651fc542bf600f0433/desc/io/equilibrium_io.py#L79-L87

my point was that in general we should probably avoid special casing that relies on particular data types, and instead declare particular attributes static or dynamic if the default behavior isn't correct. 

IE, our code shouldn't care whether something is a python `int` or `np.int64`, we should make sure it works for either.

Reply:
FWIW I think pretty much all the changes in this PR could be replaced by just adding the following to `_unjitable`:

```python
if hasattr(x, "dtype") and np.isscalar(x):
    return np.issubdtype(x.dtype, np.bool_) or np.issubdtype(x.dtype, np.int_)
```

Reply:
Minimal example on `master`, where `coilset_sym` is from the `DummyCoilSet` in `conftest`: 
```
objective = ObjectiveFunction(
    (
        CoilLength(coils=coilset_sym),
        CoilCurvature(coil=coilset_sym),
    ),
)
constraints = (FixCoilCurrent(coilset_sym),)
optimizer = Optimizer("lsq-exact")
[coilset_sym], _ = optimizer.optimize(
    things=coilset_sym,
    objective=objective,
    constraints=constraints,
    verbose=3,
)
```
Gives the error message: `AttributeError: 'numpy.bool_' object has no attribute 'N'`

With only your suggested change to `_unjitable`, it gives a different error message: 
`ValueError: Mismatch custom node data: [('_NFP', 4), ('_sym', ('ndarray', (), (True,))), ('_name', '')] != [('_NFP', 4), ('_name', '')]; value: CoilSet at 0x7fcfa9022690 (name=, with 3 submembers).`

I agree my solution in this PR is a little hacky, but it works and we need to fix this bug ASAP for coil optimizations to work. I doubt python and numpy are going to be adding new data types anytime soon that we need to add logic for. If we come up with a more elegant solution later we can always update it.

Reply:
I think that's because of the stuff like this:
```python
is_leaf=lambda x: not hasattr(x, "__len__")
```

in the coil objectives. I'm not sure what thats actually supposed to be checking but it treats numpy arrays as leafs which im pretty sure we don't want.

Reply:
Actually nevermind, I had a mistake above. the part added to unjitable should be 
```python
    if hasattr(x, "dtype") and np.ndim(x)==0:
        return np.issubdtype(x.dtype, np.bool_) or np.issubdtype(x.dtype, np.int_)
```

with that fix  on master your example above seems to run fine